backend/FwLite/MiniLcm/SyncHelpers/DiffCollection.cs

```diff
@@ -1,141 +1,267 @@
+        {
+            if (stable.Contains(current[j].Id))
             {
-                await add(api, entry);
-                return 1;
-            },
-            async (api, entry) =>
+                beforeEntity = current[j];
+                break;
+            }
+        }
+        for (var j = i + 1; j < current.Count; j++)
+        {
+            if (stable.Contains(current[j].Id))
+            {
+                afterEntity = current[j];
+                break;
+            }
+        }
+        return new BetweenPosition
+        {
+            Previous = beforeEntity?.Id,
+            Next = afterEntity?.Id
+        };
+    }
+
+    private static ImmutableSortedDictionary<int, PositionDiff>? DiffPositions<T>(
+        IList<T> before,
+        IList<T> after) where T : IOrderable
+    {
+        var beforeJson = new JsonArray([.. before.Select(item => JsonValue.Create(item.Id))]);
+        var afterJson = new JsonArray([.. after.Select(item => JsonValue.Create(item.Id))]);
+        return JsonDiffPatcher.Diff(beforeJson, afterJson, DiffFormatter.Instance);
+    }
+
+#pragma warning disable IDE0072 // Add missing cases
+    /// <summary>
+    /// Formats Json array diffs into a list sorted by:
+    /// - deletes first (with arbitrary negative indexes)
+    /// - added and moved in order of new/current index (using current index)
+    /// </summary>
+    private class DiffFormatter : IJsonDiffDeltaFormatter<ImmutableSortedDictionary<int, PositionDiff>>
+    {
+        public static readonly DiffFormatter Instance = new();
+
+        private DiffFormatter() { }
+
+        public ImmutableSortedDictionary<int, PositionDiff>? Format(ref JsonDiffDelta delta, JsonNode? left)
+        {
+            if (delta.Kind == DeltaKind.None) return null;
+
+            return delta.GetArrayChangeEnumerable().Select(change =>
             {
-                await remove(api, entry);
-                return 1;
-            },
-            async (api, beforeEntry, afterEntry) => await replace(api, beforeEntry, afterEntry));
+                return change.Diff.Kind switch
+                {
+                    DeltaKind.ArrayMove => new PositionDiff(change.Diff.GetNewIndex(), PositionDiffKind.Move),
+                    DeltaKind.Added => new PositionDiff(change.Index, PositionDiffKind.Add),
+                    DeltaKind.Deleted => new PositionDiff(change.Index, PositionDiffKind.Remove),
+                    _ => throw new InvalidOperationException("Only array diff results are supported"),
+                };
+            }).ToImmutableSortedDictionary(diff => diff.SortIndex, diff => diff);
+        }
+    }
+#pragma warning restore IDE0072 // Add missing cases
+}
+
+public enum PositionDiffKind { Add, Remove, Move }
+public record PositionDiff(int Index, PositionDiffKind Kind)
+{
+    public int SortIndex => Kind == PositionDiffKind.Remove ? -Index - 1 : Index;
+}
+
+public class BetweenPosition : IEquatable<BetweenPosition>
```

Comment:
I must have forgot to mention this before, this is exactly the use case for a record, records implement value based equality like you have defined here.